README.md


# Fringe.Py

[![GitHub version](https://badge.fury.io/gh/farhadnkm%2FFringe.Py.svg)](https://badge.fury.io/gh/farhadnkm%2FFringe.Py)
[![PyPI version](https://badge.fury.io/py/fringe.svg)](https://badge.fury.io/py/fringe)
[![DOI](https://zenodo.org/badge/327240077.svg)](https://zenodo.org/badge/latestdoi/327240077)
[![MIT License](https://img.shields.io/apm/l/atomic-design-ui.svg?)](https://github.com/farhadnkm/Fringe.Py/blob/master/LICENSE)

A coherent propagation and diffraction simulation tool, all written in Python. 

In this package, a set of utilities are provided to simulate coherent signal propagation. It is particularly made for free-space optical propagation, diffraction, and holography. However, the tools are compatible with 1D and 2D data structures and can be potentially used for any sort of spatially-coherent signals.

-----

<p align="center">
    <img src="examples/images/exports/propagation_anim.gif" width="500">
<p>


## What's inside?

A set of modules are provided to solve the scalar diffraction problem. The codes are GPU-friendly, compatible with well-known computation libraries (Numpy, TensorFlow) and they support batch processing. Using the TensorFlow backend, computations become autograd-compatible and could be mounted on machine learning models. Angular spectrum algorithm is the primary work horse for field propagation though other custom solvers could be used. Aside the built-in Numpy and TensorFlow backends, any computational means could be employed to process tensor operations.

It also includes:
- a simple yet useful data pipeline to load and standardize data. For now, it only supports images.
- Gerchberg-Saxton multi-distance phase recovery algorithm. It can be easily tweaked to support other variations of signal e.g. by wavelength.

## Installation
To install the package, run:

```
python -m pip install fringe
```

Fringe requires ```numpy```, ```tensorflow 2.x```, and ```scikit_image```.
The example files are not included in the package and should be downloaded separately. Also they require ```matplotlib``` to show plots.

## How to Use
1. Import or create data

For images:

```python
import numpy as np
from fringe.utils.io import import_image
from fringe.utils.modifiers import ImageToArray, Normalize, MakeComplex
```

Images need to be standardized, normalized, and casted to complex data type. *Modifiers* are tools made for this purpose which apply these operations on import.

```python
p1 = ImageToArray(bit_depth=16, channel='gray', crop_window=None, dtype='float32')
p2 = Normalize(background=np.ones((512, 512)))
p3 = MakeComplex(set_as='amplitude', phase=0)

obj = import_image("images/squares.png", preprocessor=[p1, p2, p3])
```

2. Propagate

*Solvers* contain propagation algorithms and can be called by *solver.solve*. In particular, angular Spectrum algorithm convolves the input field with a free-space propagtor function which depends on *wavelength λ* (or *wavenumber k=2π/λ*) and distance *z*.

```python
from numpy import pi, abs, angle
from fringe.solvers.AngularSpectrum import AngularSpectrumSolver as AsSolver

solver = AsSolver(shape=obj.shape, dr=1, is_batched=False, padding="same", pad_fill_value=0, backend="Numpy")
rec = solver.solve(hologram, k=2*pi/500e-3, z=-1000)
amp, phase = abs(rec), angle(rec)

ax = pyplot.sublots(2, 1, 1)
ax[0].imshow(abs(obj))
ax[1].imshow(amp)
pyplot.show()
```

<p align="center">
    <img src="examples/images/exports/squares_1.png" width="300"> <img src="examples/images/exports/squares_2.png" width="300">
<p>

Example notebooks provide further details with 1D and 2D diffraction, GPU acceleration, batch processing, and phase recovery.

## License

Fringe is released under the MIT license. See LICENSE for details.


HISTORY.rst
=======
History
=======

0.0.1 (2021-02-19)
------------------
* First release on PyPI.

0.0.2 (2021-02-25)
------------------
* Some bugs and errors fixed

0.0.3 (2021-02-25)
------------------
* Minor bugs fixed

0.0.5 (2023-03-15)
------------------
* AngularSpectrum.py
    Propagator issue for small pixels is resolved
    Near field propagation is now supported
* io.py
    skimage warnings io is fixed.

0.0.6 (2023-03-16)
------------------
* AngularSpectrum.py
    Bug fixed

0.0.7 (2023-03-16)
------------------
* AngularSpectrum.py
    Positive real exponential bug fixed


MHPR_TF.py
import numpy as np
import os
from skimage.restoration import unwrap_phase
from Utils.Processing_GPU import AngularSpectrumSolver as AsSolver, MultiHeightPhaseRecovery as MHPR
from Utils.DataManager import import_image, import_image_seq, export_image, convert_to_tensor_seq
from Utils.Modifiers import ImageToArray, PreprocessHologram
import matplotlib
matplotlib.use("TkAgg")
from matplotlib import pyplot as plt


bg_path = "PATH_TO_IMAGE.tif"

img_seq_paths = [
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/0_300.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/1_350.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/2_400.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/3_450.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/4_500.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/5_550.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/6_600.tif",
	"D:/Research data/results/images/selected/Simulation/test 1/MHPR/Holograms/7_650.tif"
]


pp1 = ImageToArray(bit_depth=16, channel='gray', crop_window=None, dtype='float32')
#bg = import_image(bg_path, preprocessor=pp1)
bg = np.ones((512, 512))  # no background
pp2 = PreprocessHologram(background=bg)
h_seq = import_image_seq(img_seq_paths, preprocessor=[pp1, pp2])
h_seq = convert_to_tensor_seq(h_seq, dtype='complex64')


solver = AsSolver(shape=h_seq[0].shape, dx=1.12, dy=1.12, wavelength=532e-3, dtype_f='float32', dtype_c='complex64')
z_values = [300, 350, 400, 450, 500, 550, 600, 650]

mhpr = MHPR(solver)
recovered_h = mhpr.resolve(h_seq, z_values, 100)
amp = np.abs(recovered_h)
phase = unwrap_phase(np.angle(recovered_h))
phase += np.pi
phase /= 2 * np.pi


cmap = matplotlib.cm.get_cmap('viridis')

export_image(amp, os.path.join('D:/Research data/results/images/selected/Simulation/test 1/MHPR/', 'amp.png'), dtype='uint8')
export_image(phase, os.path.join('D:/Research data/results/images/selected/Simulation/test 1/MHPR/', 'ph.png'), dtype='uint8')
export_image(cmap(phase), os.path.join('D:/Research data/results/images/selected/Simulation/test 1/MHPR/', 'ph_cm.png'), dtype='uint8')


fig, (ax1, ax2) = plt.subplots(1, 2)
ax1.imshow(amp, cmap='gray')
ax2.imshow(phase, cmap='viridis')
plt.show()
